### ISSUE TITLE
Unicode error with Python 3 when running pyftsubset on Linux

### ISSUE THREAD

**Issue by KyleAMathews** [Munter/subfont#45]
Building this Gatsby site works on my mac but is failing in our Linux (Docker) container.

If this is an upstream problem e.g. in `fonttools`, I'll move the issue over there but thought I'd start here.

This is our Dockerfile if that helps:

```dockerfile
FROM node:carbon

RUN apt-get update && \
    apt-get install -y \
        python3 \
        python3-pip \
        python3-setuptools \
        groff \
        less \
    && pip3 install --upgrade pip \
    && apt-get clean

# Add Python dependencies for subfont
RUN pip3 install fonttools brotli zopfli
```

Console output:

```
node_modules/.bin/subfont -i --no-recursive --inline-css public/index.html
Guessing --root from input files: file:///usr/src/app/www/www/public/
 ✔ 0.042 secs: logEvents
 ✔ 0.761 secs: loadAssets
 ✔ 1.227 secs: populate
 ✔ 0.174 secs: checkIncompatibleTypes
 ✔ 0.050 secs: populate
 ✔ 0.058 secs: populate
 ⚠ WARN: public/subfont/Futura_PT__subset__subset-400-c2439033a3.woff2: Command failed: pyftsubset /tmp/input-11884-22204-ggiea2.f6nxh.woff2 --output-file=/tmp/output-11884-22204-mfcpt8.q8rvf.woff2 --obfuscate_names --flavor=woff2 --text=" !&'()+,-.012348:;?ABCDEFGHIJKLMNOPQRSTUVWYabcdefghijklmnopqrstuvwxyz ·–—’“”…"
         Traceback (most recent call last):
           File "/usr/local/bin/pyftsubset", line 11, in <module>
             sys.exit(main())
           File "/usr/local/lib/python3.4/dist-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
             return func(*args, **kwds)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 3290, in main
             subsetter.populate(glyphs=glyphs, gids=gids, unicodes=unicodes, text=text)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 2888, in populate
             text_utf32 = text.encode("utf-32-be")
         UnicodeEncodeError: 'utf-32-be' codec can't encode character '\udcc2' in position 70: surrogates not allowed

 ⚠ WARN: public/subfont/Futura_PT__subset__subset-700-0ddc10d20b.woff2: Command failed: pyftsubset /tmp/input-11884-22204-xcwcrw.4r6df.woff2 --output-file=/tmp/output-11884-22204-101pg4b.yk33.woff2 --obfuscate_names --flavor=woff2 --text=" -?ABCDFGHJLMNPRSUWYabcdefghiklmnoprstuvwxyz "
         Traceback (most recent call last):
           File "/usr/local/bin/pyftsubset", line 11, in <module>
             sys.exit(main())
           File "/usr/local/lib/python3.4/dist-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
             return func(*args, **kwds)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 3290, in main
             subsetter.populate(glyphs=glyphs, gids=gids, unicodes=unicodes, text=text)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 2888, in populate
             text_utf32 = text.encode("utf-32-be")
         UnicodeEncodeError: 'utf-32-be' codec can't encode character '\udcc2' in position 45: surrogates not allowed

 ⚠ WARN: public/subfont/Futura_PT__subset__subset-400-c2439033a3.woff2: Command failed: pyftsubset /tmp/input-11884-22204-nth2tk.v7snh.woff --output-file=/tmp/output-11884-22204-s995tq.5sv3p.woff --obfuscate_names --flavor=woff --text=" !&'()+,-.012348:;?ABCDEFGHIJKLMNOPQRSTUVWYabcdefghijklmnopqrstuvwxyz ·–—’“”…" --with-zopfli
         Traceback (most recent call last):
           File "/usr/local/bin/pyftsubset", line 11, in <module>
             sys.exit(main())
           File "/usr/local/lib/python3.4/dist-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
             return func(*args, **kwds)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 3290, in main
             subsetter.populate(glyphs=glyphs, gids=gids, unicodes=unicodes, text=text)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 2888, in populate
             text_utf32 = text.encode("utf-32-be")
         UnicodeEncodeError: 'utf-32-be' codec can't encode character '\udcc2' in position 70: surrogates not allowed

 ⚠ WARN: public/subfont/Futura_PT__subset__subset-700-0ddc10d20b.woff2: Command failed: pyftsubset /tmp/input-11884-22204-74jgt4.qfbyl.woff --output-file=/tmp/output-11884-22204-1bateq7.4bu2.woff --obfuscate_names --flavor=woff --text=" -?ABCDFGHJLMNPRSUWYabcdefghiklmnoprstuvwxyz " --with-zopfli
         Traceback (most recent call last):
           File "/usr/local/bin/pyftsubset", line 11, in <module>
             sys.exit(main())
           File "/usr/local/lib/python3.4/dist-packages/fontTools/misc/loggingTools.py", line 372, in wrapper
             return func(*args, **kwds)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 3290, in main
             subsetter.populate(glyphs=glyphs, gids=gids, unicodes=unicodes, text=text)
           File "/usr/local/lib/python3.4/dist-packages/fontTools/subset/__init__.py", line 2888, in populate
             text_utf32 = text.encode("utf-32-be")
         UnicodeEncodeError: 'utf-32-be' codec can't encode character '\udcc2' in position 45: surrogates not allowed

 ✔ 3.214 secs: subsetFonts
 ✔ 0.011 secs: writeAssetsToDisc
TypeError: Expected a finite number, got number: NaN
    at module.exports (/usr/src/app/www/www/node_modules/subfont/node_modules/pretty-bytes/index.js:34:9)
    at /usr/src/app/www/www/node_modules/subfont/lib/cli.js:177:38
    at <anonymous>
```

**Comment by papandreou** [comment]
@KyleAMathews, interesting! Since `U+DCC2` is the "Low surrogate" character, I'm guessing that's not the actual offending char. Are you able to share a zip with the content of `/usr/src/app/www/www/public/`?

**Comment by KyleAMathews** [comment]
This is for https://next.gatsbyjs.org/

We have `--no-recursive` set and are only targeting the home page (for now) so it's whatever is in the initial index.html page there.

**Comment by papandreou** [comment]
Weird, it seems to work for me on Ubuntu 18.04 with Python 2:

```
$ python --version
Python 2.7.15rc1
$ pip show fonttools | grep Version
Version: 3.19.0
$ bringhome https://next.gatsbyjs.org/ -o gatsby
 ✔ 0.002 secs: logEvents
 ✔ 1.860 secs: loadAssets
 ⚠ WARN: https://storage.googleapis.com/workbox-cdn/releases/3.4.1/packages/workbox-sw/browser.mjs: HTTP 403 Forbidden
         Including assets:
             https://storage.googleapis.com/workbox-cdn/releases/3.4.1/workbox-sw.js.map
         
 ✔ 5.679 secs: populate
 ✔ 0.231 secs: checkIncompatibleTypes
 ✔ 1.219 secs: moveAssets
 ✔ 0.201 secs: writeAssetsToDisc
Output written to file:///home/andreas/work/subfont/gatsby/
$ subfont -i --no-recursive gatsby/index.html 
Guessing --root from input files: file:///home/andreas/work/subfont/gatsby/
 ✔ 0.001 secs: logEvents
 ✔ 0.659 secs: loadAssets
 ⚠ WARN: https://storage.googleapis.com/workbox-cdn/releases/3.4.1/packages/workbox-sw/browser.mjs: HTTP 403 Forbidden
         Including assets:
             gatsby/storage.googleapis.com/workbox-cdn/releases/3.4.1/workbox-sw.js.map
         
 ✔ 1.628 secs: populate
 ✔ 0.134 secs: checkIncompatibleTypes
 ✔ 0.035 secs: populate
 ✔ 0.050 secs: populate
 ⚠ WARN: inline JavaScript in gatsby/index.html: Parse error in inline JavaScript in gatsby/index.html
         Line 1: Unexpected identifier (line 1)
 ✔ 3.203 secs: subsetFonts
 ✔ 0.027 secs: writeAssetsToDisc
gatsby/index.html: 1 font (3 variants) in use, 60.8 kB total. Created subsets: 17.8 kB total
  Futura PT:
    400 : 77/230 codepoints used, 19.5 kB (woff2) => 8.33 kB (woff2)
    700 : 47/230 codepoints used, 20.5 kB (woff2) => 6.68 kB (woff2)
    400i:  8/230 codepoints used, 20.8 kB (woff2) => 2.77 kB (woff2)
HTML/JS/CSS size increase: 1.71 kB
Total savings: 41.4 kB
Output written to file:///home/andreas/work/subfont/gatsby/
```

I've tried adding some debug output to see which `pyftsubset` commands get called, and I can tell that it's creating the same subsets as the commands that fail for you :confused:

**Comment by KyleAMathews** [comment]
Yeah, it could be a Python 3 problem. I didn't try it (yet) with Python 2.

**Comment by papandreou** [comment]
I don't quite know what I'm doing when it comes to python, but I can't reproduce the problem after installing `fonttools`, `brotli` and `zopfli` packages with `pip3` and plowing a `#!/usr/bin/python3` shebang into `/home/andreas/.local/bin/pyftsubset`.

**Comment by papandreou** [comment]
WIP for getting rid of `fonttools` here: https://github.com/Munter/subfont/pull/56

**Comment by papandreou** [comment]
Fixed in 6.0.0 (which uses harfbuzz instead of fonttools)
